src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Validation.cs

```diff
@@ -432,10 +448,19 @@ private static void ShowValidationAdornerHelper(DependencyObject targetElement,
                 {
                     if (tryAgain)
                     {
-                        // try again later, perhaps giving layout a chance to create the adorner layer
-                        adornerSite.Dispatcher.BeginInvoke(DispatcherPriority.Loaded,
-                                    new DispatcherOperationCallback(ShowValidationAdornerOperation),
-                                    new object[]{targetElement, adornerSite, show});
+                        // Check if the element is visible, if not try to show the adorner again once it gets visible.
+                        // This is needed because controls hosted in Expander or TabControl don't have a parent/AdornerLayer till the Expander is expanded or the TabItem is selected.
+                        if (siteUIElement.IsVisible == false)
+                        {
+                            siteUIElement.IsVisibleChanged += ShowValidationAdornerWhenAdornerSiteGetsVisible;
```

Comment:
Can the element not become IsVisible = true before the loading finishes? Because then the repeated attempt would fail and you would not get the Loaded attempt which you get now otherwise.

Reply:
I don't know how a not loaded control should be visible.

Reply:
@miloush were you able to get any instances where the above condition will hold true? Or if you can suggest a way to check that?

Reply:
Also what if the control never gets visible? Aren't we leaking references by subscribing to IsVisibleChanged?

Reply:
We won't leak as the handler is static and not an instance member, but i will also test that.

Will have a look at the sample you posted.

Reply:
RenderTargetBitmap might be another scenario that could be affected.

Since the code already protects against re-entrancy, why don't we just keep the current codepath _and_ add your visibility check after?

Reply:
The basic premise is that `IsVisible` is set to true for all controls before the `Loaded` events are fired (and the controls first displayed) and that is still before the dispatcher gets to the Loaded priority operations. That is a lot of opportunities in between for people to mess up with the tree.

Reply:
@dipeshmsft @miloush I made changes that should prevent the code from introducing any regressions.
I also confirmed that the changes don't lead to leaks if an element never gets visible.